src/Microsoft.OData.Client/BulkUpdateSaveResult.cs

```diff
@@ -202,19 +306,370 @@ protected override ODataRequestMessageWrapper CreateRequestMessage(string method
+                        response.StatusCode,
+                        () => this.ResponseStream);
+
+                    ODataMaterializerContext materializerContext = new ODataMaterializerContext(responseInfo, base.MaterializerCache);
+                    deltaFeed = ODataReaderEntityMaterializer.ParseDeltaResourceSetPayload(responseMessageWrapper, responseInfo, entityDescriptor.Entity.GetType(), materializerContext);
+                }
+                catch (Exception ex)
+                {
+                    exception = ex;
+
+                    if (!CommonUtil.IsCatchableExceptionType(ex))
+                    {
+                        throw;
+                    }
+                }
+            }
+
+            // Cache the response.
+            this.cachedResponse = new CachedResponse(
+                this.bulkUpdateGraph,
+                headers,
+                (HttpStatusCode)response.StatusCode,
+                responseVersion,
+                deltaFeed,
+                exception);
+        }
+
+        /// <summary>
+        /// Gets the last descriptor that was processed or visited. 
+        /// </summary>
+        /// <param name="adjacentDescriptors">Tracks the descriptors adjacent to a descriptor that has been visited.</param>
+        /// <returns>The descriptor for the particular operation being processed.</returns>
+        private static Descriptor GetLastVisitedDescriptor(Stack<(int LastVisited, IReadOnlyList<Descriptor> Descriptors)> adjacentDescriptors)
+        {
+            Descriptor lastVisitedDescriptor = null;
+
+            if (adjacentDescriptors.Count > 0)
+            {
+                int index = adjacentDescriptors.Peek().LastVisited;
+
+                IReadOnlyList<Descriptor> descriptors = adjacentDescriptors.Peek().Descriptors;
+
+                if (index < descriptors.Count)
+                {
+                    lastVisitedDescriptor = descriptors[index];
+                }
+            }
+           
+            return lastVisitedDescriptor;
+        }
+
+        /// <summary>
+        /// Checks whether an entry has a data modification exception instance annotation.
+        /// </summary>
+        /// <param name="materializerState">The materializer state entry to check for the data modification exception instance annotation.</param>
+        /// <returns>true if the entry has the data modification exception, otherwise false.</returns>
+        private static bool HasFailedOperation(IMaterializerState materializerState)
+        {
+            if (materializerState is MaterializerDeletedEntry deltaEntry)
+            {
+                return deltaEntry.DeletedResource.InstanceAnnotations.Any(a => a.Name == XmlConstants.DataModificationException);
+            }
+            else if (materializerState is MaterializerEntry entry)
+            {
+                return entry.Entry.InstanceAnnotations.Any(a => a.Name == XmlConstants.DataModificationException);

```

Comment:
Is there a test where you mock a response from the service that gives rise to this scenario?

Reply:
https://github.com/OData/odata.net/blob/20122a766a87190e9cec72127d772d7de39b219c/test/FunctionalTests/Tests/DataServices/UnitTests/Client.TDD.Tests/Tests/BulkUpdateE2ETests.cs#L330